host/src/FunctionsCustomHost/AppLoader/AppLoader.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Runtime.InteropServices;
+
+namespace FunctionsCustomHost
+{
+    // If having problems with the managed host, enable the following:
+    // Environment.SetEnvironmentVariable("COREHOST_TRACE", "1");
+    // In Unix environment, you need to run the below command in the terminal to set the environment variable.
+    // export COREHOST_TRACE=1
+
+    /// <summary>
+    /// Manages loading hostfxr & worker assembly.
+    /// </summary>
+    internal sealed class AppLoader : IDisposable
+    {
+        private IntPtr _hostfxrHandle = IntPtr.Zero;
+        private IntPtr _hostContextHandle = IntPtr.Zero;
+        private bool _disposed;
+
+        internal AppLoader()
+        {
+        }
+
+        internal int RunApplication(string? assemblyPath)
+        {
+            ArgumentNullException.ThrowIfNull(assemblyPath, nameof(assemblyPath));
+
+            unsafe
+            {
+                var parameters = new NetHost.get_hostfxr_parameters
+                {
+                    size = sizeof(NetHost.get_hostfxr_parameters),
+                    assembly_path = GetCharArrayPointer(assemblyPath)
+                };
+
+                var hostfxrFullPath = NetHost.GetHostFxrPath(&parameters);
+                Logger.Log($"hostfxr path:{hostfxrFullPath}");
+
+                _hostfxrHandle = NativeLibrary.Load(hostfxrFullPath);
+
+                if (_hostfxrHandle == IntPtr.Zero)
+                {
+                    Logger.Log($"Failed to load hostfxr. hostfxrFullPath:{hostfxrFullPath}");
+                    return -1;
+                }
+
+                Logger.Log($"hostfxr loaded.");
+
+                Logger.Log($"hostfxr initialized with {assemblyPath}");
+                HostFxr.SetAppContextData(_hostContextHandle, "AZURE_FUNCTIONS_NATIVE_HOST", "1");
+
+                return HostFxr.Run(_hostContextHandle);
```

Comment:
I think we are missing the code to set `_hostContextHandle`. Looks like Initialize call is missing.

Reply:
added initialize call